### PR TITLE
plugins/ipam: add IPv6 SLAAC plugin

### PR DESCRIPTION
**Pull request by dcbw**
@containernetworking/cni-maintainers @squeed @danehans

### CODE REVIEW

**Comment by dcbw** [comment]
@squeed updated with suggestions from the maintainer meeting.  Though my memory was fuzzy, did we agree on just turning off RA once we had the details?

File: plugins/ipam/slaac/slaac.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+			return fmt.Errorf("failed to acquire IPv6 address")
+		}
+
+		// Look for an IPv6 default route through the container interface,
+		// from which we grab the gateway
+		routes, err := netlink.RouteList(link, netlink.FAMILY_V6)
+		if err != nil {
+			return fmt.Errorf("failed to list %q IPv6 routes: %v", link.Attrs().Name, err)
+		}
+		for _, r := range routes {
```
Comment (squeed) [inline plugins/ipam/slaac/slaac.go]: Do we want to grab any further prefixes? Maybe look for any other routes with a protocol of ra (`RTPROT_RA = 9`)?

Comment (squeed) [inline plugins/ipam/slaac/slaac.go]: Duh, nevermind, they'll be on-link prefixes and therefore have a corresponding SLAAC address.

Honestly, I should probably take up llama farming.

**Comment by squeed** [comment]
Tested this with a local radvd, got this:
```
failed to add IP addr {Version:6 Interface:2 Address:{IP:2001:470:6d:7bf:8054:98ff:febf:ca38 Mask:ffffffffffffffff0000000000000000} Gateway:fe80::f0d3:13ff:fe73:8c86} to "eth0": file exists
```

**Comment by dcbw** [comment]
> Tested this with a local radvd, got this:

@squeed I guess we turn of RA, remove the addresses and routes, and then pass them back?  To replace an existing address you have to use netlink.AddrReplace() instead of AddrAdd().  Perhaps we should just change ConfigureIface() to use AddrReplace() instead, since it'll add the address if not present, and replace if present?

**Comment by squeed** [comment]
That should work, I would think. One thing to be careful for is whether or not the addresses and routes still have a lifetime.

BTW, I managed to get it to work by - this is extreme - disabling ipv6, disabling slaac, then re-enabling ipv6. I actually have a bridge running radvd with the same prefix as my access network, and I currently have an online container with v6 'net access and no nat! However, something's not right, and the container can't reach other hosts on my local 'net.

**Comment by dcbw** [comment]
@squeed did the AddrReplace/RouteReplace thing

**Comment by squeed** [comment]
reviving this PR.

AddrReplace didn't fix it: `failed to add IP addr {Version:6 Interface:0xc420011388 Address:{IP:2001:470:6d:7bf:d098:e5ff:febb:bae9 Mask:ffffffffffffffff0000000000000000} Gateway:fe80::5cc8:daff:fec8:7e77} to "eth0": file exists`

**Comment by squeed** [comment]
Reviving this PR.

@dcbw I still get the same error: `failed to add IP addr {Version:6 Interface:0xc420011388 Address:{IP:2001:470:6d:7bf:d098:e5ff:febb:bae9 Mask:ffffffffffffffff0000000000000000} Gateway:fe80::5cc8:daff:fec8:7e77} to "eth0": file exists`

**Comment by benjumanji** [comment]
What does thi pr need to push it along?

**Comment by dcbw** [comment]
FWIW, this is blocked because CNI (and some runtimes that use it) have no way to asynchronously indicate changes in the IP addressing back. eg, if a new RA is received that changes the prefix, or DNS info, or routes, that is not able to be pushed back to the runtime, leading to the container state and the runtime state disagreeing. Not good.

**Comment by squeed** [comment]
It's not so disastrous :-).

This PR does SLACC *once*, then disables it on the container interface. That means that we don't necessarily need to support any notion of dynamic addresses.

This just needs to be revived and any bugs squashed

**Pull request closed without merging**